### ISSUE TITLE
Rule Change: support computing modified cyclomatic complexity

### ISSUE THREAD

**Issue by dpashk-figma** [eslint/eslint#18885]
### What rule do you want to change?

complexity

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
function a(x) {
    switch (x) {
        case 1:
            1;
            break;
        case 2:
            2;
            break;
        case 3:
            3;
            break;
        default:
            4;
    }
}
```

### What does the rule currently do for this code?

The cyclomatic complexity of the above block is **3**.

### What will the rule do after it's changed?

The **modified** cyclomatic** complexity of the above block is **1**.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

Modified cyclomatic complexity is almost exactly the same as the classic [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity), except that a switch statement only increases the value by 1 regardless of how many case statements it contains. The classic cyclomatic complexity, on the other hand, increases the complexity number for each case statement it finds (treating them effectively like separate if statements).

A number of other code analysis tools implement this metric:
* https://github.com/terryyin/lizard?tab=readme-ov-file#options
* [Understand](https://support.scitools.com/support/solutions/articles/70000582297-understanding-mccabe-cyclomatic-complexity)
* [Matlab's checkcode](https://www.mathworks.com/help/matlab/matlab_prog/measure-code-complexity-using-cyclomatic-complexity.html)
* [CodeSonar](https://codesecure.com/learn/code-complexity/)

The change is very straightforward to implement, just a few lines:

```diff
       WhileStatement: increaseComplexity,
       DoWhileStatement: increaseComplexity,
       AssignmentPattern: increaseComplexity,
-
-      // Avoid `default`
-      'SwitchCase[test]': increaseComplexity,
+      SwitchStatement: increaseComplexity,

       // Logical assignment operators have short-circuiting behavior
```

Plus some extra option handling code (as we would of course want to put this behind a new option). I'm happy to make a PR if you'll accept it.

**Comment by dpashk-figma** [comment]
My rationale for adopting modified cyclomatic complexity vs the classic one is that switch statements are often appropriate constructs for patterns such as action matching (think redux), route matching, message matching, and code maintainers may not want to move away from them. Example:

```js
switch (action.type) {
    case 'todos/todoAdded': 
      // ...
      break
    case 'todos/todoToggled': 
      // ...
      break

   // more cases...

    default:
      return state
  }
```

Of course it _is_ possible to replace the above construct with something else (a map of action handlers, a builder pattern, or using some macros/decorators/metaprogramming) but that may not always be preferred.

**Comment by amareshsm** [comment]
If a switch statement is used without break statements, cases will fall through to the next case. For example:

```
function a(x) {
    switch (x) {
        case 1:
            // no break
        case 2:
            // no break
        case 3:
            // no break
        default:
            4;
    }
}
```
In this case, if x is 1, it will fall through all cases and eventually execute the default case. Treating this switch statement as having a cyclomatic complexity of 1 might not capture the additional complexity introduced by the fall-through behaviour right? 🤔

**Comment by nzakas** [comment]
@amareshsm the fall through behavior isn't part of the calculation of complexity. You might be confusing complexity with code paths.

**Comment by dpashk-figma** [comment]
Right. Breaks in the switch statement do not affect the rule in its current version either.

**Comment by dpashk-figma** [comment]
Here's the PR: https://github.com/eslint/eslint/pull/18896

**Comment by mdjermanovic** [comment]
We have never accepted options for this rule since it is intended to work per the definition of cyclomatic complexity, and the advice was always to create a custom rule for any variations or other metrics. For example, this was the same request for switch statements: https://github.com/eslint/eslint/issues/15001

**Comment by nzakas** [comment]
@mdjermanovic I think the difference between this request and #15001 is that the latter was represented as a personal preference, and we didn't want to allow people to start arbitrarily changing how complexity was calculated. This request notes that there is a known alternative called "modified cyclomatic complexity" that changes how `switch` is calculated, so this is actually an option to cyclomatic complexity in the real world. That's why I think it makes sense to add as an option.

**Comment by mdjermanovic** [comment]
I agree with the arguments presented. Marking as accepted.

**Comment by TheThirdRace** [comment]
> [@mdjermanovic](https://github.com/mdjermanovic) I think the difference between this request and [#15001](https://github.com/eslint/eslint/issues/15001) is that the latter was represented as a personal preference, and we didn't want to allow people to start arbitrarily changing how complexity was calculated. This request notes that there is a known alternative called "modified cyclomatic complexity" that changes how switch is calculated, so this is actually an option to cyclomatic complexity in the real world. That's why I think it makes sense to add as an option.

Not expecting anything from my comment, just wanted to share some food for thoughts...

I think it's very noble to want to keep this rule as pure as possible. I understand what the nature of the rule is and why the eslint team is reluctant in accepting options for the rule.

To play the devil's advocate here, rules are nice but if they make our job overly difficult or force us to use less than ideal code, then we'll just deactivate them.

Switch statements, default values on function parameters, and optional chaining are all adding to complexity. There's no denying this, they add to complexity in the purest sense of the calculation. But on the other hand, the great majority of the time they make the code so much easier to read, they lower the "visual" code complexity for us humans. 

I would argue that the visual code complexity is much more important for maintenance of the code than the purest calculation is. Or maybe we need a second rule that cater to humans, people will decide which one they want to use. I believe the complexity rule current implementation is catering more to the "computer logic" than helping developers keep their code clean and easy to maintain. And if a rule is not helping us, we simply deactivate it... Then what's the point of the rule if people just deactivate it?

Again, appreciate all the work you do, simply wanted to give some food for thoughts 🙏
